### PR TITLE
Support BuildKit 1.4 syntax (heredocs, `--link`)

### PR DESCRIPTION
**Pull request by MisterDA**
I just saw that [BuildKit 1.4 syntax](https://github.com/moby/buildkit/blob/master/frontend/dockerfile/docs/syntax.md#here-documents) has support for here-documents in `RUN` and `COPY` instructions. I've added support for `COPY`, and used it in the small shell-scripts we output.
I've also added a new `link` flag to the `COPY` and `ADD` instructions.
The "breaking change" is that the generated Dockerfile from `gen_opam2_distro` now uses BuildKit syntax.
BuildKit syntax is only supported with Linux containers.

### CODE REVIEW

File: src-opam/dockerfile_opam.ml
Comment (maiste) [inline src-opam/dockerfile_opam.ml:104]: IMHO, this part of the syntax is harder to read than the original version.  The `{| |}` introduces complexity by mixing ocaml and bash syntax instead of the `run "cmd"`.

Comment (MisterDA) [reply src-opam/dockerfile_opam.ml:104]: Thank you for the review Etienne. I'd say the complexity is having to follow the file line-by-line through multiple `RUN` and `echo >` calls and that the code is more "readable", or at least that the file isn't being split and stands out in the OCaml as well as in the generated Dockerfile.
Since you don't share that opinion, but also that I think we can agree it's a minor change without much technical advantages, I'm going to split the commit using heredocs (even though it was the feature that motivated this change) from this PR so that maybe it can move forward.

File: src/dockerfile.mli
Comment (maiste) [inline src/dockerfile.mli:162]: Maybe here add the  `?link` with a small explanation about what it does?

**Review by maiste — APPROVED** [review]
Thanks for the split in two PRs! As you said, the syntax is an opinionated subject, so I'll comply with the majority :+1: 
The PR  LGTM! You can add the `?link` remark if you want to or not (as it also appears in the buildkit documentation).
However, I don't have the merge right on `ocaml-dockerfile` so I can't merge it....

File: src/dockerfile.mli
Comment (tmcgilchrist) [reply src/dockerfile.mli:162]: Agree, please document what `?link:bool` does.

File: CHANGES.md
Comment (tmcgilchrist) [inline CHANGES.md:43]: This will need to move to `0.8.1 unreleased` depending on how impactful the breaking change is.

File: src/dockerfile.ml
Comment (tmcgilchrist) [inline src/dockerfile.ml:145]: Please exhaustively match here.

**Review by tmcgilchrist — APPROVED** [review]
Please address the 2 remaining comments before merge.

**Comment by MisterDA** [comment]
> Please address the 2 remaining comments before merge.

Done, thanks!

**Review by tmcgilchrist — APPROVED** [review]

**Pull request merged**
